components/sup/src/manager/service.rs

```diff
@@ -931,7 +946,7 @@ impl Service {
             self.initialize();
             if self.initialized {
                 self.start(launcher, executor);
-                self.post_run();
+                self.post_run(executor);
```

Comment:
:thinking: I do not think I fully understand the purpose of `busy_services`. It makes sense that stopping a service would require special logic. We want to make sure we completely stop before starting again. It seems like the purpose of `busy_services` is to avoid race conditions, and I am having trouble coming up with race conditions with `post-run`.  Not saying they don't exist. I just cannot come up with a concrete example.

I completely agree that we need a top-level `Service` future. Right now we are executing futures at two levels: in the individual services and in the manager. This causes quite a bit of pain. For example, it would be nice to use the `wrap_async_service_operation` for the `post-run` hook to add it to the `busy_services`, but that is defined at the manager level whereas we run `post-run` at the service level. Another example is the need for the `needs_restart` flag.

Reply:
I think you're probably correct that `post-run` doesn't require extra bookkeeping... I thought about it some more, and I wasn't able to come up with a concrete example myself.